common.rkt

```diff
@@ -23,50 +25,152 @@
 
 ;; States
 (define empty-sub '())
+(define empty-diseq '())
+(define empty-types '())
+
 (define (walk t sub)
   (let ((xt (and (var? t) (assf (lambda (x) (var=? t x)) sub))))
     (if xt (walk (cdr xt) sub) t)))
+
 (define (occurs? x t sub)
   (cond ((pair? t) (or (occurs? x (walk (car t) sub) sub)
                        (occurs? x (walk (cdr t) sub) sub)))
         ((var? t)  (var=? x t))
         (else      #f)))
+
+(define (walk-types t types)
+  (cond
+    ((number? t) number?)
+    ((symbol? t) symbol?)
+    ((string? t) string?)
+    ((var? t) (let* ((xt (assf (lambda (x) (var=? t x)) types)))
+                (if xt (cdr xt) #f)))
```

Comment:
Still no need to `walk-types` on arbitrary terms.  You only need a way to reference a variable's type.

Also, there is no walking involved since there's no recursion.  It might be better to call this operation something like `var-type-ref`.